### PR TITLE
Improve Deployment Setup

### PR DESCRIPTION
**Pull request by PSchmiedmayer**
# Improve Deployment Setup

## :gear: Release Notes 
- Improve Deployment Setup

### Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: fastlane/SnapshotHelper.swift [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+        #else
+
+            guard self.app != nil else {
+                NSLog("XCUIApplication is not set. Please call setupSnapshot(app) before snapshot().")
+                return
+            }
+
+            let screenshot = XCUIScreen.main.screenshot()
+            #if os(iOS) && !targetEnvironment(macCatalyst)
+            let image = XCUIDevice.shared.orientation.isLandscape ?  fixLandscapeOrientation(image: screenshot.image) : screenshot.image
```
Comment (nriedman) [inline fastlane/SnapshotHelper.swift:174]:
```suggestion
            let image = XCUIDevice.shared.orientation.isLandscape ? fixLandscapeOrientation(image: screenshot.image) : screenshot.image
```

File: fastlane/SnapshotHelper.swift [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+            #else
+            let image = screenshot.image
+            #endif
+
+            guard var simulator = ProcessInfo().environment["SIMULATOR_DEVICE_NAME"], let screenshotsDir = screenshotsDirectory else { return }
```
Comment (nriedman) [inline fastlane/SnapshotHelper.swift:179]:
```suggestion
            guard var simulator = ProcessInfo().environment["SIMULATOR_DEVICE_NAME"], let screenshotsDir = screenshotsDirectory else {
                return
            }
```

File: fastlane/SnapshotHelper.swift [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+            }
+            let homeDir = URL(fileURLWithPath: simulatorHostHome)
+            return homeDir.appendingPathComponent(cachePath)
+        #else
+            throw SnapshotError.cannotRunOnPhysicalDevice
+        #endif
+    }
+}
+
+private extension XCUIElementAttributes {
```
Comment (nriedman) [inline fastlane/SnapshotHelper.swift:251]:
```suggestion
extension XCUIElementAttributes {
```

File: fastlane/SnapshotHelper.swift [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+    var isNetworkLoadingIndicator: Bool {
```
Comment (nriedman) [inline fastlane/SnapshotHelper.swift:252]:
```suggestion
    private var isNetworkLoadingIndicator: Bool {
```

File: fastlane/SnapshotHelper.swift [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+        if hasAllowListedIdentifier { return false }
+
+        let hasOldLoadingIndicatorSize = frame.size == CGSize(width: 10, height: 20)
+        let hasNewLoadingIndicatorSize = frame.size.width.isBetween(46, and: 47) && frame.size.height.isBetween(2, and: 3)
+
+        return hasOldLoadingIndicatorSize || hasNewLoadingIndicatorSize
+    }
+
+    var hasAllowListedIdentifier: Bool {
```
Comment (nriedman) [inline fastlane/SnapshotHelper.swift:261]:
```suggestion
    private var hasAllowListedIdentifier: Bool {
```

File: fastlane/SnapshotHelper.swift [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+        let allowListedIdentifiers = ["GeofenceLocationTrackingOn", "StandardLocationTrackingOn"]
+
+        return allowListedIdentifiers.contains(identifier)
+    }
+
+    func isStatusBar(_ deviceWidth: CGFloat) -> Bool {
```
Comment (nriedman) [inline fastlane/SnapshotHelper.swift:267]:
```suggestion
    private func isStatusBar(_ deviceWidth: CGFloat) -> Bool {
```

File: fastlane/SnapshotHelper.swift [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+        guard frame.origin == .zero else { return false }
+
+        let oldStatusBarSize = CGSize(width: deviceWidth, height: 20)
+        let newStatusBarSize = CGSize(width: deviceWidth, height: 44)
+
+        return [oldStatusBarSize, newStatusBarSize].contains(frame.size)
+    }
+}
+
+private extension XCUIElementQuery {
```
Comment (nriedman) [inline fastlane/SnapshotHelper.swift:278]: SwiftLint violation -- avoid using access controlled extensions.

File: fastlane/SnapshotHelper.swift [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+    }
+
+    var hasAllowListedIdentifier: Bool {
+        let allowListedIdentifiers = ["GeofenceLocationTrackingOn", "StandardLocationTrackingOn"]
+
+        return allowListedIdentifiers.contains(identifier)
+    }
+
+    func isStatusBar(_ deviceWidth: CGFloat) -> Bool {
+        if elementType == .statusBar { return true }
```
Comment (nriedman) [inline fastlane/SnapshotHelper.swift:268]:
```suggestion
        if elementType == .statusBar {
            return true
        }
```

File: fastlane/SnapshotHelper.swift [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+        guard frame.origin == .zero else { return false }
```
Comment (nriedman) [inline fastlane/SnapshotHelper.swift:269]:
```suggestion
        guard frame.origin == .zero else {
            return false
        }
```

File: fastlane/SnapshotHelper.swift [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+        let newStatusBarSize = CGSize(width: deviceWidth, height: 44)
+
+        return [oldStatusBarSize, newStatusBarSize].contains(frame.size)
+    }
+}
+
+private extension XCUIElementQuery {
+    var networkLoadingIndicators: XCUIElementQuery {
+        let isNetworkLoadingIndicator = NSPredicate { (evaluatedObject, _) in
+            guard let element = evaluatedObject as? XCUIElementAttributes else { return false }
```
Comment (nriedman) [inline fastlane/SnapshotHelper.swift:281]:
```suggestion
            guard let element = evaluatedObject as? XCUIElementAttributes else {
                return false
            }
```

File: fastlane/SnapshotHelper.swift [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+
+            return element.isNetworkLoadingIndicator
+        }
+
+        return self.containing(isNetworkLoadingIndicator)
+    }
+
+    @MainActor
+    var deviceStatusBars: XCUIElementQuery {
```
Comment (nriedman) [inline fastlane/SnapshotHelper.swift:290]:
```suggestion
    @MainActor var deviceStatusBars: XCUIElementQuery {
```

File: fastlane/SnapshotHelper.swift [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+        let isNetworkLoadingIndicator = NSPredicate { (evaluatedObject, _) in
+            guard let element = evaluatedObject as? XCUIElementAttributes else { return false }
+
+            return element.isNetworkLoadingIndicator
+        }
+
+        return self.containing(isNetworkLoadingIndicator)
+    }
+
+    @MainActor
```
Comment (nriedman) [inline fastlane/SnapshotHelper.swift:289]:
```suggestion
```

File: fastlane/SnapshotHelper.swift [hunk 13]
```diff
@@ -0,0 +1,315 @@
...
+    var deviceStatusBars: XCUIElementQuery {
+        guard let app = Snapshot.app else {
+            fatalError("XCUIApplication is not set. Please call setupSnapshot(app) before snapshot().")
+        }
+
+        let deviceWidth = app.windows.firstMatch.frame.width
+
+        let isStatusBar = NSPredicate { (evaluatedObject, _) in
+            guard let element = evaluatedObject as? XCUIElementAttributes else { return false }
```
Comment (nriedman) [inline fastlane/SnapshotHelper.swift:298]:
```suggestion
            guard let element = evaluatedObject as? XCUIElementAttributes else {
                return false
            }
```

File: fastlane/SnapshotHelper.swift [hunk 14]
```diff
@@ -0,0 +1,315 @@
...
+            return element.isStatusBar(deviceWidth)
+        }
+
+        return self.containing(isStatusBar)
+    }
+}
+
+private extension CGFloat {
+    func isBetween(_ numberA: CGFloat, and numberB: CGFloat) -> Bool {
+        return numberA...numberB ~= self
```
Comment (nriedman) [inline fastlane/SnapshotHelper.swift:309]:
```suggestion
       numberA...numberB ~= self
```

**Review by nriedman — APPROVED** [review]
I went through and identified some changes that would prevent SwiftLint from throwing errors.

**Comment by PSchmiedmayer** [comment]
@nriedman We had a short Slack conversation about this PR. As you can see in the GitHub Action logs, updating some of the dependencies and fixing smaller bugs is currently causing some UI tests to fail. It would be amazing if you can take a look at the current set of tests and see why they are currently failing and fix them so we can merge the PR.

@Supereg While investigating some login issues, I have encountered that the updated Account + Firebase Module no longer reliably populates the account details in the `AccountSetup { details in` closure:
https://github.com/StanfordBDHG/ENGAGE-HF-iOS/blob/9784905b97f4ef19bb0bb65bb1f7b386cf05c26b/ENGAGEHF/Onboarding/AccountOnboarding.swift#L21-L26
The check above always results in the InvitationCodeView being shown despite the fact that an invitation account is attached to the user account when logging in with an existing user account. I suspect that there is a concurrent mismatch that is not waiting until the Account Storage provider returns the initial setup of user details or that they are not populated before the closure is called. I think it would make sense to continue to show the loading state until the details are loaded (and maybe show some state in the UI that the user account details are loaded).

At this point, I added the following workaround in the Invitation Code view that continuously checks if the invitation code is loaded and then proceeds with the onboarding flow:
https://github.com/StanfordBDHG/ENGAGE-HF-iOS/blob/9784905b97f4ef19bb0bb65bb1f7b386cf05c26b/ENGAGEHF/Onboarding/InvitationCodeView.swift#L69-L91

You can replicate the issue using the local emulator. You can clone the repo and log in with one of the provided accounts by first seeding the Firebase Emulator & then running the emulator.
```sh
npm --prefix ENGAGE-HF-Firebase run prepare && firebase emulators:exec -c ./ENGAGE-HF-Firebase/firebase.json --export-on-exit=./firebase 'npm --prefix ./ENGAGE-HF-Firebase/functions run serve:seed'
firebase emulators:start -c ./ENGAGE-HF-Firebase/firebase.json --import ./firebase
```

If we can guarantee that the account details are loaded before the closure is called, we could remove that workaround and directly navigate to the correct screen in the next step. Let me know what you think about the issue @Supereg 🚀

**Comment by Supereg** [comment]
@PSchmiedmayer https://github.com/StanfordSpezi/SpeziAccount/pull/79 fixes the issue. As described in the PR description, AccountServices did always call the storage provider "in sync" before submitting account details. However, a new requirement with SpeziAccount 2.0 is that StorageProvider must return immediately and perform the loading step asynchronously (e.g., the concept of "incomplete" account details exists). This wasn't checked for in the `AccountSetup` view. We added this little check and a new unit test checking that behavior.

I verified in this branch removing some of the code that this resolved the issue 👍 I left this branch for now, as you know best what was all part of the workaround and what know. But let me know if I should invest time to update this branch 👍

File: ENGAGEHF.xcodeproj/project.pbxproj [hunk 15]
```diff
@@ -2091,8 +2091,8 @@
 			isa = XCRemoteSwiftPackageReference;
 			repositoryURL = "https://github.com/StanfordSpezi/SpeziFirebase.git";
 			requirement = {
-				kind = upToNextMajorVersion;
-				minimumVersion = 2.0.0;
+				branch = "fix/storage-never-returns";
```
Comment (PSchmiedmayer) [inline ENGAGEHF.xcodeproj/project.pbxproj:2094]: TODO to address this before the merge.

**Comment by PSchmiedmayer** [comment]
Thank you for the fix @Supereg! Amazing to see all the tests passing now 🚀

**Pull request merged**